src/pykeen/training/training_loop.py

```diff
@@ -673,3 +742,62 @@ def _free_graph_and_cache(self):
         self.model.regularizer.reset()
         # The cache of the previous run has to be freed to allow accurate memory availability estimates
         torch.cuda.empty_cache()
+
+    def _save_state(self, path: str, stopper: Optional[Stopper] = None) -> None:
+        """Save the state of the training loop.
+
+        :param path:
+            Path of the file where to store the state in.
+        :param stopper:
+            An instance of :class:`pykeen.stopper.EarlyStopper` with settings for checking
+            if training should stop early
+        """
+        logger.debug("=> Saving checkpoint.")
+
+        if stopper is None:
+            stopper_dict = dict()
+        else:
+            stopper_dict = stopper.get_summary_dict()
+
+        torch.save(
+            {
+                'epoch': self._epoch,
+                'loss': self.losses_per_epochs,
+                'model_state_dict': self.model.state_dict(),
+                'optimizer_state_dict': self.optimizer.state_dict(),
+                'checksum': self.checksum,
+                'stopper_dict': stopper_dict,
+            },
+            path,
+        )
+        logger.info(f"=> Saved checkpoint after having finished epoch {self._epoch}.")
+
+    def _load_state(self, path: str) -> Mapping[str, Any]:
+        """Load the state of the training loop from a checkpoint.
+
+        :param path:
+            Path of the file where to load the state from.
+
+        :return:
+            The summary dict of the stopper at the time of saving the checkpoint.
+
+        :raises FileExistsError:
+            If the given checkpoint file has a non-matching checksum, i.e. it was saved with a different configuration.
+        """
+        logger.info(f"=> loading checkpoint '{path}'")
+        checkpoint = torch.load(path)
+        loaded_checksum = checkpoint['checksum']
+        if loaded_checksum == self.checksum:
+            self._epoch = checkpoint['epoch']
+            self.losses_per_epochs = checkpoint['loss']
+            self.model.load_state_dict(checkpoint['model_state_dict'])
+            self.optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
+            stopper_dict = checkpoint['stopper_dict']
+            logger.info(f"=> loaded checkpoint '{path}' stopped after having finished epoch {checkpoint['epoch']}")
+        else:
+            raise FileExistsError(
```

Comment:
I am not sure whether `FileExistsError` is appropriate here considering its [documentation](https://docs.python.org/3/library/exceptions.html#FileExistsError).

Reply:
Do you have a better suggestions? 😄

Reply:
I did not really find anything fitting [here](https://docs.python.org/3/library/exceptions.html). It should be something like an exception raised for corrupt files, or bad input.

Maybe we can define our own exception?

Reply:
I'll look into it.

Reply:
`ValueError` or `RuntimeError` are fine. If you want, you can subclass one/both of them to give a more specific name

Reply:
I added the CheckpointMismatchError to the training loop.